### ISSUE TITLE
New feature - Support external Open ID Connect provider

### ISSUE THREAD

**Issue by fastlorenzo** [Mailu/Mailu#2575]
### Environment

- [ x] docker-compose
- [ x] kubernetes
- [ x] docker swarm

### Version

- Version: `master`

## Description

It would be great to be able to use external authentication systems using Open ID Connect (Keycloak in my case).
I stumbled upon a [fork that does that](https://github.com/heviat/Mailu-OpenID/). I've tried to reach out to the author to identify if they wanted to make a PR to the main repo, but as I still got no answer decided to go ahead and propose the changes here directly.
All credits goes to them for the implementation.

I've tested this for a couple of weeks now and everything looks stable. I've isolated the changes required in PR #2574.

## Replication Steps

N/A

## Observed behaviour

N/A

## Expected behaviour

N/A

## Logs

N/A

**Comment by smolinari** [comment]
Commenting to be able to follow this PR, as I don't trust the :eyes: emoji. :grin: 

Scott

**Comment by nextgens** [comment]
As discussed on the PR and #mailu-dev, the way forward for this is to document how to do it with a 3rd party tool through the #1972 functionality.

Whether https://github.com/oauth2-proxy/oauth2-proxy or something else is what ends up being documented is still open for discussion; so is extending what is already in Mailu so that attributes can be passed through by the IdP.

**Comment by batistein** [comment]
highly appreciated feature! the use via oauth2 proxy sounds extremely cumbersome to me. We use this tool for kubernetes and I dislike it there as well. I would definitely favor having native support in a application. For many organizations this would be a deal breaker. 
I also tested the fork on the weekend and so far no issues, please see also: https://github.com/heviat/Mailu-OpenID/issues/1#issuecomment-1426220099

Both forks are currently working!

**Comment by nextgens** [comment]
Comments like this make me wonder why I bother to contribute to OSS.

You have forks that do provide the features you want, use them.

Have you reported upstream the reasons why you dislike oauth2-proxy or are you just whining here?

"For many organizations", who are you talking about here? What deal are you talking about? Them using Mailu? Who cares and why should anyone?

**Comment by migs35323** [comment]
nice! looking forward for this feature!

**Comment by konradmoesch** [comment]
Seeing that this feature request has more than 40 upvotes, are there any plans on integrating native OICD support into mailu? 
I really like mailu, but this is the biggest missing feature for me. I'd really like to get native support (not via some external proxy) for OICD in mailu
I am willing to help and contribute, but am concerned that any PR on this will be closed again, and I will waste my time for this..

**Comment by brunoscota** [comment]
Most wanted feature!!!! 🙌

**Comment by nextgens** [comment]
> Would this just be for the Web Interface or does this feature imply a more holistic integration with SASL XOAUTH2 as well?

SASL XOAUTH2 is discussed on #783 ; Currently there is no progress as the IETF working group has not come up with a solution yet, see https://datatracker.ietf.org/doc/draft-jenkins-oauth-public/ and https://datatracker.ietf.org/doc/draft-ietf-mailmaint-autoconfig/ if you care about progress. They have killed https://datatracker.ietf.org/doc/draft-bucksch-mauth/ recently and that is a sign that they will most likely settle for an over-engineered piece of poorly specified spec that no one will implement in an interoperable way.

Mailu won't implement something that no MUA support; MUAs wait for RFC to land.

@sierja 
> I am willing to help and contribute

Start by speccing it out. OAUTH2 is a buzzword that people throw around a lot but I am still unclear on what exactly is expected.
Making front an IdP will eventually happen (for SASL XOAUTH2 if nothing else). Making front a client to other IdPs is already possible either with the proxy-auth workaround or native integration done in forks. It remains unclear to me what a native integration would provide that the workaround does not... and this ticket stops short of explaining it.

If you do want it to happen start there: explain in plain english what it would enable (that is not currently possible).

**Comment by NicolasCaous** [comment]
@nextgens Hi, let me give you my perspective as a new user of Mailu:

> It remains unclear to me what a native integration would provide that the workaround does not... and this ticket stops short of explaining it.

Some people have different deployment infrastruture that makes managing an OpenID proxy like you suggested a bit of a headache. See the example of our coleague that runs mailu inside kubernetes, where there is already a workflow for authentication that would have to be monkey-patched just for mailu. Having native OIDC/OAuth2 authentication would make mailu better to manage for organizations that already have a OIDC/OAuth2 infrastructure.

And, to be fair, it is quite unusual for an application to not have native OIDC/OAuth2 and at the same time support it in a weird proxy deployment scheme. Almost all projects that "want" to support OIDC/OAuth2 try to make it easy to set it up... And that would be the case for mailu as well... If we could have automatic OIDC/OAuth2 by just changing mailu.env, it would be nice, like it is in most OSS software that supports OIDC/OAuth2... You just set some env vars or edit an admin page and boom, OIDC/OAuth2 works.

The point here is convenience, not that it is not possible. There is no point in making a feature like OIDC/OAuth2 hard to use if native support could be implemented.

I want to end by saying that I want to thank you for your contributions. I mean well, I just wanted to share my perspective as new user.

Let me know if you want to discuss further, my goal here is to convince you that this feature is worthwhile.

Edit: Also, I forgot to say that using a fork is a bad option for people because you don't get updates and support from the project maintainers... It is a gamble that 99% of the time is not worth it.

Edit2: Missed opportunity:

> If you do want it to happen start there: explain in plain english what it would enable (that is not currently possible).

Have some environment variables that allow OIDC to be easily configured in mailu.env. Currently this involves a complex deployment and would simplify things for users that want to adopt OIDC.

There is also the option to implement a new page in the admin view to configure it dynamically, like Gitea does:

<img width="2540" height="1115" alt="Image" src="https://github.com/user-attachments/assets/2ed63c56-650d-4301-9b48-a8cc5adfb8ff" />

<img width="2097" height="1198" alt="Image" src="https://github.com/user-attachments/assets/150746e1-fe0f-4d2f-b71a-ae46257cfd73" />

<img width="510" height="424" alt="Image" src="https://github.com/user-attachments/assets/930b4f33-a703-48f4-88e1-19906fac5405" />
